qualtran/cirq_interop/_cirq_interop.py

```diff
@@ -111,67 +167,85 @@ def cirq_optree_to_cbloq(
     If `signature` is not None, the signature of the resultant CompositeBloq is `signature`. For
     multi-dimensional registers and registers with > 1 bitsize, this function automatically
     splits the input soquets into a flat list and joins the output soquets into the correct shape
-    to ensure compatibility with the flat API expected by Cirq.
+    to ensure compatibility with the flat API expected by Cirq. When specifying a signature, users
+    must also specify the `cirq_quregs` argument, which is a mapping of cirq qubits used in the
+    OP-TREE corresponding to the `signature`. If `signature` has registers with entry
+        - `Register('x', bitsize=2, shape=(3, 4))` and
+        - `Register('y', bitsize=1, shape=(10, 20))`
+    then `cirq_quregs` should one entry corresponding to each register as follows:
+        - key='x'; value=`np.array(cirq_qubits_used_in_optree, shape=(3, 4, 2))` and
+        - key='y'; value=`np.array(cirq_qubits_used_in_optree, shape=(10, 20))`.
+
 
     If `signature` is None, the resultant composite bloq will have one thru-register named "qubits"
     of shape `(n_qubits,)`.
     """
-    # "qubits" means cirq qubits | "qvars" means bloq Soquets
     circuit = cirq.Circuit(optree)
-    all_qubits = sorted(circuit.all_qubits())
+    # "qubits" means cirq qubits | "qvars" means bloq Soquets
     if signature is None:
+        assert cirq_quregs is None
+        all_qubits = sorted(circuit.all_qubits())
         signature = Signature([Register('qubits', 1, shape=(len(all_qubits),))])
+        cirq_quregs = {'qubits': np.array(all_qubits).reshape(len(all_qubits), 1)}
+
+    assert signature is not None and cirq_quregs is not None
```

Comment:
can you tell me (in this PR) what used to happen if you provided signature and not `cirq_quregs` (because that argument didn't exist) and why you'd like to change that behavior by requiring explicit cirq_quregs

Reply:
When converting a cirq op-tree to Bloqs, we need to have a mapping between qubits used in the cirq op-tree to the corresponding named soquets in the Bloqs. Prior to this PR, the mapping was created by creating a flat list of qubits (via `all_qubits = sorted(circuit.all_qubits())`) and a flat list of soquets (by iterating over the left registers in `signature` and flattening out every multi-dimensional collection of soquets) and then asserting that the two lists are of the same length. 

As you can imagine, this brute force would work only for the most trivial cases and quickly breaks down the 1:1 mapping between the corresponding sorted lists does not hold true. 

Now, forcing users to provide the cirq qubits in the `cirq_quregs` format enables us to wire up the `CirqGateAsBloq`s correctly and note that this information already exists at all the call sites. The trivial case, where the signature of composite bloq is a single register called `qubits` of shape `(len(all_qubits),)` and `bitsize=1` is still supported. 

See the tests that I've added for more context.